### PR TITLE
compute_ctl: Add endpoint that allows setting role grants

### PR DESCRIPTION
**Pull request by devjv**
This PR introduces a `/grants` endpoint which allows setting specific `privileges` to certain `role` for a certain `schema`. 

Related to #9344 

Together these endpoints will be used to configure JWT extension and set correct usage to its schema to specific roles that will need them.

### CODE REVIEW

File: compute_tools/src/http/api.rs [hunk 1]
```diff
@@ -165,6 +169,58 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
...
+                return Response::new(Body::from(msg));
+            }
+
+            let request = hyper::body::to_bytes(req.into_body()).await.unwrap();
+            let request = serde_json::from_slice::<SetRoleGrantsRequest>(&request).unwrap();
+
+            let privileges: Result<Vec<Privilege>, _> = request
+                .privileges
+                .iter()
+                .map(|p| Privilege::from_str(p.as_str()))
```
Comment (devjv) [inline compute_tools/src/http/api.rs:190]: I wanted some validation, is this too much effort? Could also just check that the string is in some array.

File: compute_tools/src/compute.rs [hunk 2]
```diff
@@ -1367,6 +1369,39 @@ LIMIT 100",
...
+        role_name: &str,
+    ) -> Result<()> {
+        let mut conf = Config::from_str(self.connstr.as_str()).unwrap();
+        conf.dbname(db_name);
+
+        let mut db_client = conf
+            .connect(NoTls)
+            .context("Failed to connect to the database")?;
+
+        let query = "GRANT $1 ON SCHEMA $2 TO $3";
```
Comment (conradludgate) [inline compute_tools/src/compute.rs:1386]: does this work? I have my doubts

Comment (devjv) [reply compute_tools/src/compute.rs:1386]: right, guess we need to make it just a formatted string, then

File: compute_tools/src/privilege.rs
Comment (conradludgate) [inline compute_tools/src/privilege.rs:1]: can we move this to libs/compute_api ?

File: compute_tools/src/privilege.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+use std::str::FromStr;
+
+#[derive(Debug)]
+pub enum Privilege {
```
Comment (conradludgate) [inline compute_tools/src/privilege.rs:4]: If you add:
```suggestion
#[derive(Debug, serde::Deserialize, serde::Serialize)]
#[serde(rename_all = "UPPERCASE")]
pub enum Privilege
```

then you can use `pub privileges: Vec<Privelege>` in the `SetRoleGrantsRequest` and `SetRoleGrantsResponse` structs, as opposed to parsing it manually later on.

Comment (devjv) [reply compute_tools/src/privilege.rs]: Cool, so I guess I can also remove the manual `from_str` and `as_str`. Goof stuff

Comment (devjv) [reply compute_tools/src/privilege.rs]: Does it then just fail to serialize in case of some string that is not in the enum and return a similar error as we had before?

**Comment by devjv** [comment]
FYI @tristan957 I will fix this to use async still

File: compute_tools/src/compute.rs [hunk 4]
```diff
@@ -1369,32 +1369,35 @@ LIMIT 100",
...
 
+        // todo: pg_quote
         let query = format!(
             "GRANT {} ON SCHEMA {} TO {}",
-            privileges
-                .iter()
-                .map(|p| serde_json::to_string(p).unwrap())
-                .collect::<Vec<String>>()
-                .join(", "),
+            privileges.iter().map(|p| p.as_str()).join(", "),
```
Comment (conradludgate) [inline compute_tools/src/compute.rs:1394]: I'd keep what you had for now - our changes diverged a bit 😅

File: compute_tools/src/compute.rs [hunk 5]
```diff
@@ -1367,6 +1368,46 @@ LIMIT 100",
...
 
+    pub async fn set_role_grants(
+        &self,
+        db_name: &str,
+        schema_name: &str,
+        privileges: &[Privilege],
+        role_name: &str,
+    ) -> Result<()> {
+        use tokio_postgres::config::Config;
+        use tokio_postgres::NoTls;
```
Comment (tristan957) [inline compute_tools/src/compute.rs:1379]: Any reason you didn't import these at the top-level? This is not a pattern I've seen in the code.

File: compute_tools/src/compute.rs [hunk 6]
```diff
@@ -1367,6 +1368,46 @@ LIMIT 100",
...
+        conf.dbname(db_name);
+
+        let (db_client, conn) = conf
+            .connect(NoTls)
+            .await
+            .context("Failed to connect to the database")?;
+        tokio::spawn(conn);
+
+        let query = format!(
+            "GRANT {} ON SCHEMA {} TO {}",
```
Comment (tristan957) [inline compute_tools/src/compute.rs:1391]: I think making this schema specific is fine for now and this endpoint can grow other GRANT types in the future.

File: compute_tools/src/http/api.rs [hunk 7]
```diff
@@ -165,6 +167,46 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
             }
         }
 
+        (&Method::POST, "/grants") => {
+            info!("serving /grants POST request");
```
Comment (tristan957) [inline compute_tools/src/http/api.rs:171]: Similar debug comment to the other PR

File: compute_tools/src/http/api.rs [hunk 8]
```diff
@@ -165,6 +167,46 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
...
+            let status = compute.get_status();
+            if status != ComputeStatus::Running {
+                let msg = format!(
+                    "invalid compute status for set_role_grants request: {:?}",
+                    status
+                );
+                error!(msg);
+                return Response::new(Body::from(msg));
+            }
```
Comment (tristan957) [inline compute_tools/src/http/api.rs:180]: Did we have the matching check for this in the install endpoint?

File: compute_tools/src/http/openapi_spec.yaml [hunk 9]
```diff
@@ -70,7 +70,7 @@ paths:
   /info:
     get:
       tags:
-      - Info
+        - Info
```
Comment (tristan957) [inline compute_tools/src/http/openapi_spec.yaml:73]: Same unrelated code comment

File: compute_tools/src/http/openapi_spec.yaml [hunk 10]
```diff
@@ -127,10 +127,38 @@ paths:
...
+      summary: Apply grants to the database.
+      description: ""
+      operationId: setRoleGrants
+      requestBody:
+        description: Grants request.
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: SetRoleGrantsRequest
```
Comment (tristan957) [inline compute_tools/src/http/openapi_spec.yaml:143]: Does this not need match the style of the other `ref`s

**Review by tristan957 — DISMISSED** [review]

File: compute_tools/src/http/api.rs [hunk 11]
```diff
@@ -165,6 +167,46 @@ async fn routes(req: Request<Body>, compute: &Arc<ComputeNode>) -> Response<Body
...
+        (&Method::POST, "/grants") => {
+            info!("serving /grants POST request");
+            let status = compute.get_status();
+            if status != ComputeStatus::Running {
+                let msg = format!(
+                    "invalid compute status for set_role_grants request: {:?}",
+                    status
+                );
+                error!(msg);
+                return Response::new(Body::from(msg));
```
Comment (ololobus) [inline compute_tools/src/http/api.rs:179]: Here, the response is returned as plain text, but in the spec you define it as a GenericError json.

I'd suggest having a look at how `/configure` handles errors
- it returns 412 for the wrong status
- it uses `render_json_error` to render a structured json

It's a pretty small change, but will help us keeping all new APIs consistent (I know that there are some old ones that return plain text)

Comment (ololobus) [reply compute_tools/src/http/api.rs:179]: Unrelated, but @lubennikovaav look like `/installed_extensions` is inconsistent in how it returns errors, could be both plain text and json

File: libs/compute_api/src/requests.rs [hunk 12]
```diff
@@ -12,3 +12,11 @@ use serde::Deserialize;
 pub struct ConfigurationRequest {
     pub spec: ComputeSpec,
 }
+
+#[derive(Deserialize, Debug)]
+pub struct SetRoleGrantsRequest {
+    pub database: String,
```
Comment (ololobus) [inline libs/compute_api/src/requests.rs:18]: NIT: I wonder, does it make sense to use `PgIdent` here for database, schema and role?

Yes, it's currently defined as `pub type PgIdent = String;`, which as Conrad recently mentioned might not be idiomatic, but at least sends the right message that it should be escaped

File: libs/compute_api/src/responses.rs [hunk 13]
```diff
@@ -168,3 +171,11 @@ pub struct InstalledExtension {
 pub struct InstalledExtensions {
     pub extensions: Vec<InstalledExtension>,
 }
+
+#[derive(Clone, Debug, Default, Serialize)]
+pub struct SetRoleGrantsResponse {
+    pub database: String,
```
Comment (ololobus) [inline libs/compute_api/src/responses.rs:177]: NIT: same comment about `PGIdent` usage here.

File: compute_tools/src/http/openapi_spec.yaml (hunk 10 above)
Comment (ololobus) [inline compute_tools/src/http/openapi_spec.yaml:143]: Must be a full path

```suggestion
              $ref: "#/components/schemas/SetRoleGrantsRequest
```

Comment (ololobus) [reply compute_tools/src/http/openapi_spec.yaml:143]: I usually use https://editor.swagger.io/ to edit and verify the spec

File: compute_tools/src/compute.rs [hunk 14]
```diff
@@ -1367,6 +1368,46 @@ LIMIT 100",
         download_size
     }
 
+    pub async fn set_role_grants(
```
Comment (ololobus) [inline compute_tools/src/compute.rs:1371]: It'd be nice to test this flow.

It can be a very simple test:
- start compute
- create role
- add some grants via API that you expect to add in prod
- check with sql that grants were successfully added
- maybe test some edge cases, like try to add the same grants twice
- (optional) on role deletion some grants might prevent the deletion, so we have some code that reassigns owned objects and drops privileges before deletion. I think it should work, but properly it'd be great to also try to delete such role via `compute_ctl`. To do that, you need to restart compute/endpoint with slightly modified spec via `respec()` helper and by adding json
   ```json
    "delta_operations": [
        {
            "action": "delete_role",
            "name": "test_role"
        }
    ]
   ```
   I think that should work

As an example, you could reference to https://github.com/neondatabase/neon/blob/8a114e3aeda7a2e321fa4524335c1748448cae07/test_runner/regress/test_installed_extensions.py#L6 and commit that added it

**Review by ololobus — COMMENTED** [review]
I've left some comments to make code more consistent and slightly improve quality (if I didn't miss something). I'd expect it to be a few hours to address them, but let me know if it's so pressing that you cannot spend this time

After a brief look, 'it should work' (c), so we may also leave it as is and then do a follow up PR. I'd still prefer to address the comments right away

File: compute_tools/src/compute.rs (hunk 5 above)
Comment (conradludgate) [reply compute_tools/src/compute.rs:1379]: conflicts with the other import of `postgres::Config` used elsewhere in the file for non-async code.

File: compute_tools/src/http/api.rs (hunk 7 above)
Comment (conradludgate) [reply compute_tools/src/http/api.rs:171]: As discussed in the other thread, this will remain info level for now.

**Comment by devjv** [comment]
@tristan957 @ololobus feedback has been addressed

File: compute_tools/src/compute.rs [hunk 15]
```diff
@@ -1367,6 +1368,47 @@ LIMIT 100",
...
+        &self,
+        db_name: &str,
+        schema_name: &str,
+        privileges: &[Privilege],
+        role_name: &str,
```
Comment (ololobus) [inline compute_tools/src/compute.rs:1376]: NIT

```suggestion
        db_name: &PgIdent,
        schema_name: &PgIdent,
        privileges: &[Privilege],
        role_name: &PgIdent,
```
?

If it's possible and makes sense

File: compute_tools/src/http/openapi_spec.yaml [hunk 16]
```diff
@@ -127,6 +127,34 @@ paths:
...
+            schema:
+                $ref: "#/components/schemas/SetRoleGrantsRequest"
+      responses:
+        200:
+          description: Grants applied.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/SetRoleGrantsResponse"
+        500:
```
Comment (ololobus) [inline compute_tools/src/http/openapi_spec.yaml:151]: NIT

```suggestion
        412:
          description: |
            Compute is not in the right state for processing the request.
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/GenericError"
        500:
```

File: compute_tools/src/compute.rs (hunk 15 above)
Comment (ololobus) [reply compute_tools/src/compute.rs:1376]: BTW, not sure, but maybe if passed by value, it won't require `.to_string()` below?

**Review by ololobus — APPROVED** [review]
